[skip travis] Add more content to embed README.md

compulim opened the pull request.

corinagum commented:
This should be linked to in the Migration section of the root README.md

corinagum commented:
Can you add an explanation for the lay-man what Gemini is?

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+### Why I am not seeing Web Chat v4 on my page?
+
+Look at the <kbd>F12</kbd> console log and see how version selection is being done.
```
```suggestion
Look at the console log on your web page by pressing <kbd>F12</kbd> to see view the version of Web Chat.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+Look at the <kbd>F12</kbd> console log and see how version selection is being done.
+
+If you don't see the line, it is probably because the new Gemini page is not rolled out completely. If your bot is in, the URL should become https://webchat.botframework.com/embed/your-bot-id/gemini. If your bot is not in, it will not have `/gemini` appended.
```
```suggestion
If you don't see the above line, this probably means that the new Gemini page has not completely rolled out. If your bot is using Gemini, the URL will be https://webchat.botframework.com/embed/your-bot-id/gemini. If your bot is not in, it will not have `/gemini` appended.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+If you don't see the line, it is probably because the new Gemini page is not rolled out completely. If your bot is in, the URL should become https://webchat.botframework.com/embed/your-bot-id/gemini. If your bot is not in, it will not have `/gemini` appended.
+
+### I want to go back to previous versions
```
```suggestion
### I want to go back to previous versions of Web Chat
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+There are a handful of versions we support:
+
+- Majors
+   - `4` will serve the latest minor version of `4.*`. As of time of this writing, it is `4.3.0`
```
```suggestion
   - `4` will serve the latest minor version of `4.*`. At the time this document was created, that version was `4.3.0`
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+- Majors
+   - `4` will serve the latest minor version of `4.*`. As of time of this writing, it is `4.3.0`
+   - `3` will serve the latest minor version of `0.*`. As of time of this writing, it is `0.15.1-v3.748a85f`
```
```suggestion
   - `3` will serve the latest minor version of `0.*`.At the time this document was created, it is `0.15.1-v3.748a85f`
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+- Majors
+   - `4` will serve the latest minor version of `4.*`. As of time of this writing, it is `4.3.0`
+   - `3` will serve the latest minor version of `0.*`. As of time of this writing, it is `0.15.1-v3.748a85f`
+   - `scorpio` will serve the version before Gemini page is introduced. As of time of this writing, it is `0.11.4-ibiza.f373b1d`
```
```suggestion
   - `scorpio` will serve the version before Gemini page is introduced. At the time this document was created,, it is `0.11.4-ibiza.f373b1d`
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+   - `scorpio` will serve the version before Gemini page is introduced. As of time of this writing, it is `0.11.4-ibiza.f373b1d`
+      - This will be same as "Enable preview" checkbox checked
+   - `aries` will serve the version before Gemini page is introduced
+      - This will be same as "Enable preview" checkbox unchecked
```
```suggestion
      - This is the same as having the "Enable preview" checkbox unchecked
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+      - This will be same as "Enable preview" checkbox checked
+   - `aries` will serve the version before Gemini page is introduced
+      - This will be same as "Enable preview" checkbox unchecked
+      - This version will show a bot icon and welcome message on the top of the page
```
```suggestion
      - This version will show a bot icon and welcome message at the top of the page
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+      - This will be same as "Enable preview" checkbox unchecked
+      - This version will show a bot icon and welcome message on the top of the page
+- Minors
+   - `4.2` will serve the latest patch version of `4.2.*`. As of time of this writing, it is `4.2.0`
```
```suggestion
   - `4.2` will serve the latest patch version of `4.2.*`. At the time this document was created, it is `4.2.0`
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+- Minors
+   - `4.2` will serve the latest patch version of `4.2.*`. As of time of this writing, it is `4.2.0`
+
+However, we do not support specifying patch versions, e.g. `4.2.1`. This is because patch versions are for important security fixes, we will roll it out to users who are on the same minor version and guarantee security fixes will not introduce breaking changes.
```
```suggestion
However, we do not support specifying patch versions, e.g. `4.2.1`. This is because patch versions are for important security fixes. We will roll it out to users who are on the same minor version and we guarantee that security fixes will not introduce breaking changes.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+### I really don't like the versioning system
+
+We understand that some time, you will want to lock down to every single bit of code. Feel free to clone our code and build your own copy. This will make sure your bits are locked down and controlled by your system. But it also means we will not be able to serve you any new features or security fixes.
```
```suggestion
We understand that sometimes, you will want to lock down to every single bit of code. Feel free to clone our code and build your own copy. This will ensure that your bits are locked down and controlled by your system. But it also means we will not be able to serve you any new features or security fixes.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+We understand that some time, you will want to lock down to every single bit of code. Feel free to clone our code and build your own copy. This will make sure your bits are locked down and controlled by your system. But it also means we will not be able to serve you any new features or security fixes.
+
+### How staged deployment work?
```
```suggestion
### How does staged deployment work?
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+### How staged deployment work?
+
+From time to time, when we release new version of Web Chat, we will update the servicing plan in this repository and upload it to our server. It give our users transparency on when we push out a new version and how it is getting done.
```
```suggestion
From time to time, when we release a new version of Web Chat, we will update the servicing plan in this repository and upload it to our server. This gives our users transparency on when we push out a new version and how it is done.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+From time to time, when we release new version of Web Chat, we will update the servicing plan in this repository and upload it to our server. It give our users transparency on when we push out a new version and how it is getting done.
+
+Then, we gradually migrate sites to our new version using a mechanism we call dial. We turn the dial from 0% to 100%, then everyone will get the latest version. We can cherry-pick specific site into certain version, or out.
```
```suggestion
We gradually migrate customers' sites to our new version using a mechanism we call a dial. When we finish turning the dial from 0% to 100%, all customers have been switched to the latest version. We can also cherry-pick specific sites into specific version, or out.
```

compulim commented on packages/embed/README.md:
````diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+Web Chat: Selecting version "default" -> "4" -> "4.3" -> "4.3.0".
+```
+
+In the above code, the version specifier is not specified. Web Chat is using the "default" version, which forwards to major version 4, then minor version 4.3, then patch version 4.3.0.
````
```suggestion
In the above code, the version specifier is not specified. Web Chat is using the "default" version, which forwards to major version 4, then minor version 4.3, then patch version 4.3.0. You can select version by appending the URL query parameter `v` with the version number. For example, if you prefer 4.3, you should append `&v=4.3` to the URL.
```

compulim commented on packages/embed/README.md:
````diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
 ```sh
 curl https://cdn.botframework.com/botframework-webchat/4.3.0/webchat.js | openssl dgst -sha384 -binary | openssl base64 -A
````
Added.

compulim commented on packages/embed/README.md:
````diff
@@ -26,3 +26,100 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+When combined with the FiddlerScript above, "Test in Web Chat" in Azure Portal will point to the locally-developed version of the Gemini page.
+
+```js
+if (oSession.uriContains("https://webchat.botframework.com/embed/your-bot-id?features=webchatpreview&t=")) {
````
Added.

compulim commented:
@corinagum updated. Do you think this README.md can be converted into a blog post? If not, what sections do you think it missed, or what writing styles should we update it to?

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,106 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+There are a handful of versions we support:
+
+- Majors
+   - `4` will serve the latest minor version of `4.*`. At the time this document was created, that version was `4.3.0`
```
I'm assuming this is Gemini?

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +26,106 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+[This script](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity#Tools_for_generating_SRI_hashes) will fetch the asset from CDN, pipe it into OpenSSL for digest, and output as BASE64.
+
+## Test locally
```
I don't like how this article starts because the top header says "Test locally" without explaining what we're testing or what this page is about. I think we need an intro.

corinagum commented:
I think this would need more structure to be a blog post. I prefer it as a README.md just because it means we have the immediate access to update it if there are changes. A blog post may be good for announcing this new feature, but then it should link to this `README.md`.

Couple other things, I think you missed my early comments
- this page needs to have a link coming from the Migration section of the root README.md
- What is Gemini, Aries, etc? - I know what Gemini is, but does a casual Web Chat user have access to this information?
- The beginning seems abrupt to me. It just kind of jumps into code without explaining what's going on. I think this might be improved with a short introduction along the lines of "Azure Bot Services provides the method of embedding your bot into your page with `<iframe>`. This article/README outlines the the latest on using the embed code from Azure Bot Services, how to changes what version of Web Chat your client is using, and commonly asked questions related to embedding your bot."

Or something like that.

I also think the sections "What URL parameters are supported" and "I want to go back to previous versions of Web Chat" might benefit from being extracted out of FAQ and put near the 'intro' section.

compulim commented:
@corinagum I have updated the `README.md` with a couple of things:

- Added "Background" and "How to use" section
   - We could extract these two parts to become part of our blog post
- In "Background" section, there is a link to the migration guide
   - Please suggest if this link is too "deep" in the paragraph and should be digging out
   - Later on, we need to tell people which integration methods they should use
- In "FAQs" section, I have included our historical code names and their purpose
- In "Background" section, I linked "URL parameters" and "versioning"
   - Please suggest if this link is too "deep"

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
+
+- Reading configurations from Azure Bot Services
+- Simplified version selection
+   - Based on the versions you selected (latest, specific major, or specific minor), we will serve Web Chat client with latest security fixes
```
```suggestion
   - Based on the versions you selected (latest, specific major, or specific minor), we will serve the Web Chat client with latest security fixes
```

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
+- Reading configurations from Azure Bot Services
+- Simplified version selection
+   - Based on the versions you selected (latest, specific major, or specific minor), we will serve Web Chat client with latest security fixes
+   - Please refer to [this section](#i-want-to-go-back-to-previous-versions-of-web-chat) to specifying version to use
```
```suggestion
   - Please refer to the [I want to go back](#i-want-to-go-back-to-previous-versions-of-web-chat) to specify the version you want to use
```

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
+- Simplified version selection
+   - Based on the versions you selected (latest, specific major, or specific minor), we will serve Web Chat client with latest security fixes
+   - Please refer to [this section](#i-want-to-go-back-to-previous-versions-of-web-chat) to specifying version to use
+- Same experince as "Test in Web Chat" blade in Azure Portal
```
```suggestion
- This creates the same experience as "Test in Web Chat" blade in Azure Portal
```

corinagum commented on packages/embed/README.md:
````diff
@@ -1,23 +1,57 @@
+<iframe src="https://webchat.botframework.com/embed/your-bot-id?s=your-site-secret"></iframe>
+```
+
+For production use, we strongly prefer you to use token instead of secret. Please refer to [this article](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication) for how to generate token out of secret.
````
bold this the first sentence in this paragraph.

compulim replied:
Done.

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
 1. `npm run prepublishOnly`
-1. Publishes the build artifact at `/dist/index.html`, to our internal repository
+1. Publishes the build artifact at `/dist/index.html`
    1. Verifies the file size, it should be less than 30 KB
```
```suggestion
   1. Verifies the file size. It should be less than 30 KB
```

compulim replied:
Weird, but Markdown use indentation of 3 spaces.

compulim replied:
Done.

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
 
 `servicingPlan.json` is the deployment plan. It provides details about where to locate assets, how to set up, and when to deprecate our older bits.
 
 When modifying the servicing plan to point to newer assets, make sure it is pointing to immutable copies on CDN. For example, assets with hashed filenames. Since we are using subresource integrity, this will make sure servicing plan and assets can be invalidated on CDN cache asynchronously.
```
```suggestion
When modifying the servicing plan to point to newer assets, make sure it is pointing to immutable copies on CDN. For example, you want to use assets with hashed filenames. Since we are using subresource integrity, this will make sure the servicing plan and assets can be invalidated on CDN cache asynchronously.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+## FAQs
+
+Below are common questions about the new Gemini page.
```
```suggestion
Below are common questions about the new Gemini deployment of Web Chat.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+Integrating Web Chat using `<iframe>` is the simplest way to connect your website with Azure Bot Services.
+
+Although `<iframe>` do not support a wide variety of customization options, it help you to connect easier by filling out options, such as bot avatar initials, user ID, etc. We also maintain versions for you to make sure your client is up-to-date and secure all the time.
```
```suggestion
Although `<iframe>` does not support a wide variety of customization options, it helps you connect more easily by filling out options, such as bot avatar initials, user ID, etc. We also maintain versions for you to make sure your client is up-to-date and secure all the time.
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+- Majors
+   - `4` will serve the latest minor version of `4.*`. At the time this document was created, that version was `4.3.0`
+   - `3` will serve the latest minor version of `0.*`.At the time this document was created, it is `0.15.1-v3.748a85f`
+   - `scorpio` will serve the version before Gemini page is introduced. At the time this document was created,, it is `0.11.4-ibiza.f373b1d`
```
```suggestion
   - `scorpio` will serve the version before Gemini page was introduced. At the time this document was created, `scorpio` is version `0.11.4-ibiza.f373b1d`
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+   - `4` will serve the latest minor version of `4.*`. At the time this document was created, that version was `4.3.0`
+   - `3` will serve the latest minor version of `0.*`.At the time this document was created, it is `0.15.1-v3.748a85f`
+   - `scorpio` will serve the version before Gemini page is introduced. At the time this document was created,, it is `0.11.4-ibiza.f373b1d`
+      - This will be same as "Enable preview" checkbox checked
```
```suggestion
      - This is the as having the "Enable preview" checkbox checked
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+   - `3` will serve the latest minor version of `0.*`.At the time this document was created, it is `0.15.1-v3.748a85f`
+   - `scorpio` will serve the version before Gemini page is introduced. At the time this document was created,, it is `0.11.4-ibiza.f373b1d`
+      - This will be same as "Enable preview" checkbox checked
+   - `aries` will serve the version before Gemini page is introduced
```
```suggestion
   - `aries` will serve the version before Gemini page was introduced
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+
+Historically, we use Zodiac signs as code name to version the embed page.
+
+- Gemini: this version of embed page
```
```suggestion
- Gemini: the latest version of embed page
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+Historically, we use Zodiac signs as code name to version the embed page.
+
+- Gemini: this version of embed page
+   - Supports multiple versions of Web Chat, backward- and forward-compatible
```
```suggestion
   - Supports multiple versions of Web Chat, backwards- and forwards-compatible
```

corinagum commented on packages/embed/README.md:
```diff
@@ -26,3 +60,129 @@ To generate SHA384 for a specific asset from an URL, run the following command i
+   - Supports multiple versions of Web Chat, backward- and forward-compatible
+      - Latest version is maintained at [`botframework-webchat@latest`](https://www.npmjs.com/package/botframework-webchat/)
+      - v3 is maintained at [`botframework-webchat@legacy`](https://www.npmjs.com/package/botframework-webchat/v/legacy), latest is 0.15.1
+   - Supports falling back to previous versions of embed page, include Scorpio and Aries
```
```suggestion
   - Supports falling back to previous versions of embed page include Scorpio and Aries
```

compulim commented:
Resolved all. Could you review it again?

- Modified root `README.md`
- Bolded the "For production use" sentence
- Updated the "Verifies the file size." line
- Incorporated all your other reviews

corinagum commented on README.md:
```diff
@@ -12,25 +12,28 @@
 
-### My Bot's webpage is rendering Web Chat v3 and uses customization options provided by Web Chat, no customization at all, or very little of my own customization that was not available with Web Chat.
-- Please follow the implementation of sample [`01.c.getting-started-migration`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/01.c.getting-started-migration) to convert your webpage from v3 to v4 of Web Chat.
+Starting from May 2019, we are rolling out v4 to websites that integrate Web Chat using `<iframe>` element. Please refer to [this article](https://github.com/Microsoft/BotFramework-WebChat/treemaster/packages/embed/README.md) for information on integrating Web Chat using `<iframe>`.
```
```suggestion
Starting from May 2019, we are rolling out v4 to websites that integrate Web Chat using `<iframe>` element. Please refer to [the embed documentation](https://github.com/Microsoft/BotFramework-WebChat/treemaster/packages/embed/README.md) for information on integrating Web Chat using `<iframe>`.
```

Don't forget that saying 'this documentation' is not good accessibility, since it doesn't tell the user what they will be clicking.

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
+- Reading configurations from Azure Bot Services
+- Simplified version selection
+   - Based on the versions you selected (latest, specific major, or specific minor), we will serve the Web Chat client with latest security fixes
+   - Please refer to the [I want to go back](#i-want-to-go-back-to-previous-versions-of-web-chat) to specify the version you want to use
```
```suggestion
   - Please refer to the [I want to go back documentation](#i-want-to-go-back-to-previous-versions-of-web-chat) to specify the version you want to use
```

corinagum commented on packages/embed/README.md:
```diff
@@ -1,23 +1,57 @@
+- Simplified version selection
+   - Based on the versions you selected (latest, specific major, or specific minor), we will serve the Web Chat client with latest security fixes
+   - Please refer to the [I want to go back](#i-want-to-go-back-to-previous-versions-of-web-chat) to specify the version you want to use
+- This creates the same experience as "Test in Web Chat" blade in Azure Portal
```
```suggestion
- This creates the same experience as the "Test in Web Chat" blade in Azure Portal
```
